src/memo/memo.ml

```diff
@@ -1250,12 +1261,13 @@ module Invalidation = struct
     | Empty -> true
     | _ -> false
 
-  let clear_caches = Leaf (fun () -> Caches.clear ())
-
-  let clear_cache { cache; _ } =
+  let clear_caches =
```

Comment:
Even though this change should help greatly, I'm still not sure that `clear_caches` does an easy to understand thing, in particular it can't guarantee that no memoized results from previous runs will be used.

For example, it still does not invalidate the lazy nodes, and it doesn't invalidate the nodes "forgotten" previously (this is an issue inherited from `invalidate_store`: what if we drop the node from the store, but it's later computed again?)